src/compiler/checker.ts

```diff
@@ -8016,6 +8154,65 @@ module ts {
             }
         }
 
+        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {
+            var seen: boolean[] = [];
+            if (!type || !globalIPromiseType || !(type.flags & TypeFlags.ObjectType)) {
+                return fallbackType;
+            }
+            seen[type.id] = true;
+            if (checkTypeRelatedTo(type, globalIPromiseType, assignableRelation, undefined)) {
+                var thenProp = getPropertyOfType(type, "then");
+                var thenType = getTypeOfSymbol(thenProp);
+                var thenSignatures = getSignaturesOfType(thenType, SignatureKind.Call);
+                var onFulfilledParameterType = getUnionType(map(thenSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                var onFulfilledParameterSignatures = getSignaturesOfType(onFulfilledParameterType, SignatureKind.Call);
+                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));

```

Comment:
This seems pretty reasonable, but beware that signatures degenerate under union if they are not exactly the same (modulo return type).